connaisseur/alert.py

```diff
@@ -0,0 +1,187 @@
+        self.throw_if_alert_sending_fails = receiver_config.get(
+            "fail_if_alert_sending_fails", False
+        )
+        self.payload = self._construct_payload(receiver_config)
+        self.headers = _get_headers(receiver_config)
+
+    def _construct_payload(self, receiver_config):
+        try:
+            alert_templates_dir = f'{os.getenv("ALERT_CONFIG_DIR")}/templates'
+            with safe_path_func(
+                open,
+                alert_templates_dir,
+                f"{alert_templates_dir}/{self.template}.json",
+                "r",
+            ) as templatefile:
+                template = json.load(templatefile)
+        except Exception as err:
+            raise ConfigurationError(
+                "Template file for alerting payload is either missing or invalid JSON: {}".format(
+                    str(err)
+                )
+            )
+        payload = self._render_template(template)
+        if receiver_config.get("payload_fields") is not None:
+            payload.update(receiver_config.get("payload_fields"))
+        return json.dumps(payload)
+
+    def _render_template(self, template_dict):
+        if isinstance(template_dict, (dict, list)):
+            for key, value in template_dict.items():
+                if isinstance(value, dict):
+                    self._render_template(value)
+                elif isinstance(value, list):
+                    for entry in value:
+                        self._render_template(entry)
+                else:
+                    template_dict[key] = Template(value).render(
+                        self.context, undefined=StrictUndefined
+                    )
+        return template_dict
+
+    def send_alert(self):
+        try:
+            response = requests.post(
+                self.receiver_url, data=self.payload, headers=self.headers
+            )
+            response.raise_for_status()
+            logging.info("sent alert to %s", self.template)
+        except Exception as err:
+            if self.throw_if_alert_sending_fails:
+                raise AlertSendingError(str(err))
+            logging.error(err)
+        return response
+
+
+def _get_headers(receiver_config):
+    headers = {"Content-Type": "application/json"}
+    additional_headers = receiver_config.get("custom_headers")
+    if additional_headers is not None:
+        for header in additional_headers:
+            key, value = header.split(":", 1)
+            headers.update({key.strip(): value.strip()})
+    return headers
+
+
+def load_config():
+    try:
+        alert_config_dir = f'{os.getenv("ALERT_CONFIG_DIR")}'
+        with safe_path_func(
+            open, alert_config_dir, f"{alert_config_dir}/alertconfig.json", "r"
+        ) as configfile:
+            alertconfig = json.load(configfile)
+            schema = get_alert_config_validation_schema()
+            json_validate(instance=alertconfig, schema=schema)
```

Comment:
I somehow can't reply to the [other conversation](https://github.com/sse-secure-systems/connaisseur/pull/70#discussion_r583754780), so here's the response: The context manager generally only should contain the code that actually needs to access the provided context (here: `configfile`). So you could unindent the other lines.